File: MATLAB/blades.m
%% Impeller Blades

%% Meanline Definition
r_inlet_blade(p) = r_inlet_impeller(p); % m - radial location of blade inlets - typically vertically aligned with eye radius - this assumption can be changed
meanline_curve_untrimmed(:,:,p) = 0.5 * (shroud_curve(:,:,p) + impeller_curve(:,:,p)); % [m,m] - halfway between shroud and impeller
inlet_index(p) = find(meanline_curve_untrimmed(:,1,p) > r_inlet_blade, 1); % unitless - point at which meanline curve lies directly beneath eye radius

r_meanline = linspace(r_inlet_blade(p), r_outlet_impeller(p), 1000)'; % m
z_meanline = pchip(meanline_curve_untrimmed(:,1,p), meanline_curve_untrimmed(:,2,p), r_meanline); % m
meanline_curve(:,:,p) = [r_meanline, z_meanline]; % [m,m] - blade meanline

% meanline_curve_blade =  meanline_curve(inlet_index(p):end,:,p); % [m,m] the portion of the meanline with a blade

%% Velocity Triangles
u_blade(:,p) = pump_shaft_speed(p)*meanline_curve(:,1,p); % m/s - blade tangential velocity
v_merid(:,p) = vdot_pump(p)/(pi*(r_inlet_impeller(p)^2-r_hub(p)^2)).*ones(1000,1); % m/s - NEEDS TO VARY ALONG MEANLINE
% v_merid(:,p) = vdot_pump(p) ./ (2*pi*meanline_curve(:,1,p).*vecnorm(shroud_curve(:,:,p)-impeller_curve(:,:,p),2,2)); % m/s - meridional velocity

hydraulic_efficiency(p) = 1 - 0.071 / vdot_pump(p)^0.25; % unitless - Jekat's Empirical Formula - valid for all specific speeds CONVERT TO ANDERSON?

v_tangential_inlet(p) = 0; % m/s - assuming no pre-swirl from the inducer
v_tangential_outlet(p) = g*head_pump(p) / (hydraulic_efficiency(p)*u_blade(end,p)) + v_tangential_inlet(p); 
    % m/s - this is Euler's pump equation rearranged (V_theta,2 in pump handbook). The tangential flow speed to achieve head requirement.                                                                                                                                                                                                                                                                                                                                                                                                                           
v_tangential(:,p) = linspace(v_tangential_inlet(p), v_tangential_outlet(p), length(meanline_curve(:,p)))'; % assuming linear ramp of fluid tangential velocity

fluid_azimuth(:,p) = atan(v_merid(:,p) ./ (u_blade(:,p) - v_tangential(:,p))); % rad - angle between fluid flow azimuth and local tangential azimuth (beta_f2 in pump handbook)

%% Camber Line Integration - Blade Curve Construction
blade_azimuth(:,p) = atan2(v_merid(:,p), (u_blade(:,p).*(1-slip_factor(:,p)) - v_tangential(:,p)));    % rad - angle between blade tip azimuth and local tangential azimuth (beta_2 in pump handbook)

dm = sqrt( diff(r_meanline).^2 + diff(z_meanline).^2 ); % meanline differential arc length
meanline_arc_length(:,p) = [0;cumsum(dm)]; % meanline cumulative arc length
theta_blade(:,p) = cumtrapz(r_meanline,cot(blade_azimuth(:,p))./r_meanline); % rad

blade_curve(:,:,p) = [meanline_curve(:,1,p).*[cos(theta_blade(:,p)),-clocking(p)*sin(theta_blade(:,p))],meanline_curve(:,2,p)]; % [m,m,m]

%% Number of Blades
blade_arc_length(p) = sum(sqrt(diff(blade_curve(:,1,p)).^2+diff(blade_curve(:,2,p)).^2+diff(blade_curve(:,3,p)).^2)); % m - arc length of one individual blade, measure at meanline line
solidity_ideal(p) = interp1([0, 0.4, 3], [1.8, 1.8, 1], specific_speed(p)); % solidity is the optimal ratio of blade chord to blade spacing. Pump handbook page 2.36 (sigma)
blade_count(p) = round(solidity_ideal(p) * 2*pi*r_outlet_impeller(p) / blade_arc_length(p)); % number of blades
solidity(p) = blade_count(p)*blade_arc_length(p)/(2*pi*r_outlet_impeller(p)); % unitless

%% Compute Blockage (pump handbook page 2.63)
blade_thickness(p) = 0.04*r_outlet_impeller(p); % m
boundary_layer_thickness(p) = 0.002*blade_arc_length(p); % m
meridional_length(p) = sum(sqrt(diff(blade_curve(:,1,p)).^2+diff(blade_curve(:,3,p)).^2)); % m - arc length of blade projected onto the meridional plane
blockage(p) = 1 - solidity(p)/meanline_arc_length(end,p)*(boundary_layer_thickness(p)+blade_thickness(p)); % unitless - proportion of area not blocked by blades

%% Velocity Slip - Pfliederer's Method - pump handbook pg 2.35
% Extending this to the full arc length of the meanline, instead of just the tip
% a = 0.75;  % unitless - 0.65 to 0.85 for volute
% r_ref = r_meanline(1); % m - scalar reference radius
% psi_prime = a * (1 + sin(blade_azimuth(:,p))); % unitless
% slip_factor(:,p) = psi_prime .* r_meanline .^ 2 ./  (blade_count(p) * 0.5 * (r_meanline.^2 - r_ref^2)); % unitless

% = fraction of impeller discharge meridional area (that is, the area
% normal to the velocity component Vm, 2) that is not blocked by the
% thickness of the blades and the boundary layer displacement
% thickness on blades and on hub and shroud surfaces

%% Cavitation Check
% eqns 25 and 26 in pump handbook

File: Python/blades.py
import numpy as np
import matplotlib.pyplot as plt
from scipy.interpolate import CubicSpline
from scipy.integrate import cumulative_trapezoid
import math
from dataclasses import dataclass
from engine_state import engine, pump

def blades(p: pump):
    imp = p.impeller[0]  # Currently only single stage pumps are supported
    b = imp.blade

    # Meanline Definition
    inlet_idx = int(0.1*imp.meanline_curve.shape[0]) # index where blades start
    b.meanline_curve = imp.meanline_curve[inlet_idx:, :] # [m,m] - the portion of the meanline with a blade
    
    #######################################################################################################
    ### Velocity triangles
    #######################################################################################################
    imp.u_meanline = p.shaft_speed * imp.meanline_curve[:, 0] # m/s - tangential velocity along the meanline
    b.u = p.shaft_speed * b.meanline_curve[:, 0] # m/s - blade tangential velocity
    imp.v_merid = p.vdot / imp.area_meanline # m/s - meridional velocity
    b.v_merid = imp.v_merid[inlet_idx:] # m/s - meridional velocity with blockage correction

    # imp.v_tangential_inlet = ind.v_meanline[-1]
    imp.v_tangential_inlet = imp.u_meanline[0] # m/s - assuming ideal swirl off inducer UPDATE WITH ACTUAL VALUE FROM INDUCER

    imp.v_tangential_outlet = engine.g * p.head_rise / (imp.hydraulic_efficiency * b.u[-1]) + imp.v_tangential_inlet # m/s
    b.v_tangential = np.linspace(imp.v_tangential_inlet, imp.v_tangential_outlet, b.meanline_curve.shape[0]) # m/s - assuming linear ramp of fluid tangential velocity
    imp.c_bladed = np.sqrt(b.v_merid**2 + b.v_tangential**2) # m/s - absolute velocity along the bladed portion of the meanline

    #######################################################################################################
    ### Camber Line Integration - Blade Curve Construction
    #######################################################################################################
    b.angle = np.arctan2(b.v_merid, (b.u * (1.0 - imp.slip_factor) - b.v_tangential)) # rad - angle between blade tip azimuth and local tangential azimuth (beta_2 in pump handbook)
    dm = np.sqrt(np.diff(b.meanline_curve[:,0])**2 + np.diff(b.meanline_curve[:,1])**2) # m - differential arc length along meanline
    imp.meanline_arc_length = np.concatenate([[0.0], np.cumsum(dm)]) # m - meanline cumulative arc length

    theta_blade = cumulative_trapezoid(1.0/np.tan(b.angle)/b.meanline_curve[:,0], b.meanline_curve[:,0], initial=0)  # rad - amgular step to advance blade curve

    imp.blade_curve = np.column_stack([b.meanline_curve[:,0]*np.cos(theta_blade),b.meanline_curve[:,0]*-p.clocking*np.sin(theta_blade),b.meanline_curve[:,1],])

    #######################################################################################################
    ### Number of Blades + Blockage
    #######################################################################################################
    b.ds = np.linalg.norm(np.diff(imp.blade_curve, axis=0), axis=1) # m
    b.arc_length = np.concatenate([[0.0], np.cumsum(b.ds)]) # m - cumulative arc length along the blade curve (length N)
    b.arc_length_total = float(b.ds.sum()) # m

    solidity_ideal = float(CubicSpline([0.0, 0.4, 3.0], [1.8, 1.8, 1.0])(imp.specific_speed))
        # unitless - solidity is the optimal ratio of blade chord to blade spacing. Pump handbook page 2.36 (sigma)
    test = solidity_ideal * (2.0 * np.pi * imp.r_outlet) / b.ds
    imp.blade_count = int(np.round(solidity_ideal * (2.0 * np.pi * imp.r_outlet) / b.arc_length_total)) # number of blades

    imp.solidity = imp.blade_count * b.arc_length_total / (2.0 * np.pi * imp.r_outlet) # unitless

    imp.blade_thickness = 0.04 * imp.r_outlet # m - empirical assumption: 4% of outlet radius as blade thickness
    imp.boundary_layer_thickness = 0.002 * b.arc_length_total # m

    imp.meridional_length = np.sum(np.sqrt(np.diff(imp.blade_curve[:, 0])**2 + np.diff(imp.blade_curve[:, 2])**2)) # m - arc length of blade projected onto the meridional plane

    imp.blockage[inlet_idx:] = 1 - imp.blade_count * (2*imp.boundary_layer_thickness + imp.blade_thickness) / (2*np.pi*b.meanline_curve[:,0]) / np.sin(b.angle) # unitless

    b.area_meanline = imp.area_meanline[inlet_idx:] # m2 - crosswise area along meanline

    #######################################################################################################
    ### Velocity Slip
    #######################################################################################################
    
    # Pfleiderer’s Method - pump handbook pg 2.35 - Extending this to the full arc length of the meanline, instead of just the tip
    a = 0.75 # unitless - 0.65 to 0.85 for volute 
    r_ref = b.meanline_curve[0,0] # m - scalar reference radius
    psi_prime = a * (1.0 + np.sin(b.angle)) # unitless
    denom = imp.blade_count * 0.5 * b.meanline_curve[:,0]**2 - r_ref**2 
    imp.slip_factor = psi_prime * (b.meanline_curve[:,0]**2) / denom # unitless

    # Cavitation Check
    # eqns 25 and 26 in pump handbook
    # add cavitation criteria at blade inlet/outlet

    #######################################################################################################
    ### Hydraulic Efficiency
    #######################################################################################################

    b.v_merid_avg = np.trapz(b.v_merid, b.arc_length) / b.arc_length_total # m/s - average meridional velocity along the blade
    b.passage_width = 2*np.pi*b.meanline_curve[:,0]*imp.blockage[inlet_idx:] / imp.blade_count # m - circumferential distance between blades at impeller outlet
    b.area_meanline_avg = np.trapezoid(b.area_meanline, b.arc_length) / b.arc_length_total # m2 - average crosswise area along the bladed portion of the meanline
    b.hydraulic_diameter_avg = np.sqrt(4/np.pi*b.area_meanline_avg) # m - hydraulic diameter of flow passage - Gulich 3.8.4
    
    reynolds = b.v_merid_avg*b.arc_length_total*p.density/p.viscosity # unitless - Gulich 3.8.3
    friction_coeff = 0.136/(-np.log10(0.2*imp.surface_roughness/b.arc_length_total+12.5/reynolds))**2.15 # unitless - friction coefficient - Gulich 3.8.3
    dissipation_coeff = (friction_coeff+0.0015)*(1.1+4*b.passage_width[-1]/(2*np.pi*imp.r_outlet)) # unitless - dissipation coefficient - Gulich 3.8.5
    friction_loss = 4*dissipation_coeff*b.arc_length_total/b.hydraulic_diameter_avg*(b.v_merid_avg/b.u[-1])**2 # unitless - friction loss coefficient - Gulich 3.8.6

    b.u_meanline = p.shaft_speed * b.meanline_curve[:,0] # m/s - tangential velocity along the meanline
    b.w_inlet = np.sqrt(b.v_merid[0]**2 + (b.u_meanline[0] - b.v_tangential[0])**2) # m/s - inlet relative velocity - Gulich 3.8.7
    if  b.v_merid[0]/b.w_inlet > 0.65:
        shock_loss = 0.3*((b.w_inlet-b.v_merid[0])/b.u[-1])**2 # unitless - shock loss coefficient - Gulich 3.8.8
    else:
        shock_loss = 0.0
    
    imp.total_loss = friction_loss + shock_loss # unitless - total loss coefficient - Gulich 3.8.10

    

